phys2bids/conversion.py

```diff
@@ -0,0 +1,403 @@
+"""
+import os.path as op
+from operator import itemgetter
+from itertools import groupby
+import logging
+import matplotlib.pyplot as plt
+
+from bids import BIDSLayout
+import pandas as pd
+import numpy as np
+import nibabel as nib
+
+from .bids import update_bids_name
+from .slice4phys import slice_phys
+from .physio_obj import BlueprintOutput
+
+LGR = logging.getLogger(__name__)
+
+
+def load_scan_data(layout, sub, ses=None):
+    """
+    Extract subject- and session-specific scan onsets and durations from BIDSLayout.
+
+    Parameters
+    ----------
+    layout : BIDSLayout
+    sub : str
+    ses : str or None, optional
+
+    Returns
+    -------
+    df : pandas.DataFrame
+        DataFrame with the following columns: 'filename', 'original_filename',
+        'acq_time', 'duration', 'onset'.
+        'filename' is the name of the BIDS file minus within-acquisition entities.
+        'original_filename' is the actual, existing BIDS file.
+        'acq_time' is the acquisition time in datetime format.
+        'duration' is the scan duration in seconds.
+        'onset' is the scan onset in seconds, starting with zero at the first scan.
+    """
+    # This is the strategy we'll use in the future. Commented out for now.
+    # scans_file = layout.get(extension='.tsv', suffix='scans', subject=sub, session=ses)
+    # df = pd.read_table(scans_file)
+
+    # Collect acquisition times
+    # NOTE: Will be replaced with scans file when heudiconv makes the change
+    img_files = layout.get(datatype='func', suffix='bold',
+                           extension=['.nii.gz', '.nii'],
+                           subject=sub, session=ses)
+    df = pd.DataFrame(
+        columns=['original_filename', 'acq_time'],
+    )
+    for i, img_file in enumerate(img_files):
+        df.loc[i, 'original_filename'] = img_file.path
+        df.loc[i, 'acq_time'] = img_file.get_metadata()['AcquisitionTime']
+
+    # Get generic filenames (without within-acquisition entities like echo)
+    df['filename'] = df['original_filename'].apply(update_bids_name,
+                                                   echo=None, fa=None,
+                                                   inv=None, mt=None, part=None,
+                                                   ch=None)
+
+    # Get "first" scan from multi-file acquisitions
+    df['acq_time'] = pd.to_datetime(df['acq_time'])
+    df = df.sort_values(by='acq_time')
+    df = df.drop_duplicates(subset='filename', keep='first', ignore_index=True)
+
+    # Now back to general-purpose code
+    df = determine_scan_durations(layout, df, sub=sub, ses=ses)
+    df = df.dropna(subset=['duration'])  # limit to relevant scans
+
+    # Convert scan times to relative onsets (first scan is at 0 seconds)
+    df['onset'] = (df['acq_time'] - df['acq_time'].min())
+    df['onset'] = df['onset'].dt.total_seconds()
+    return df
+
+
+def determine_scan_durations(layout, scan_df, sub, ses=None):
```

Comment:
@RayStick I would love to get your input on this function. It makes the very simple assumption that scan duration = number of volumes * repetition time, but I don't know if that holds for ASL. We could also load and look at [`AcquisitionDuration` and other fields](https://bids-specification.readthedocs.io/en/stable/04-modality-specific-files/01-magnetic-resonance-imaging-data.html#timing-parameters_1), if that would help.

Reply:
_Most_ of the time this is the case for ASL outputs, but sometimes it isn't. I think, for now, it's fair to assume this (duration=vols*reps), and then we work on making it more adaptable later? In the funny cases where there is one long TR at the start (i.e. M0scan or 'calibration' scan proceeding the main ASL scan) it might be simplest if the user splits these two files first but I am not fully sure - I'd need to think about it a bit more.